addons/stock/tests/test_move2.py

```diff
@@ -511,6 +511,68 @@ def test_pick_pack_ship_return(self):
+            'product_uom_qty': 4,
+            'product_uom': self.productB.uom_id.id,
+            'picking_id': picking_ship.id,
+            'location_id': self.output_location,
+            'location_dest_id': self.customer_location,
+        })
+
+        pack_move = self.env['stock.move'].create({
+            'name': self.productB.name,
+            'product_id': self.productB.id,
+            'product_uom_qty': 4,
+            'product_uom': self.productB.uom_id.id,
+            'picking_id': picking_pack.id,
+            'location_id': self.pack_location,
+            'location_dest_id': self.output_location,
+            'move_dest_ids': [(4, ship_move.id)],
+        })
+
+        self.env['stock.move'].create({
+            'name': self.productB.name,
+            'product_id': self.productB.id,
+            'product_uom_qty': 4,
+            'product_uom': self.productB.uom_id.id,
+            'picking_id': picking_pick.id,
+            'location_id': self.stock_location,
+            'location_dest_id': self.pack_location,
+            'move_dest_ids': [(4, pack_move.id)],
+            'state': 'confirmed',
+        })
+        stock_location = self.env['stock.location'].browse(self.stock_location)
+
+        self.env['stock.quant']._update_available_quantity(self.productA, stock_location, 1.0)
+        self.env['stock.quant']._update_available_quantity(self.productB, stock_location, 4.0)
+        picking_pick.action_assign()
+        picking_pick.move_line_ids.filtered(lambda ml: ml.product_id == self.productA).qty_done = 1.0
+        picking_pick.move_line_ids.filtered(lambda ml: ml.product_id == self.productB).qty_done = 2.0
+        first_pack = picking_pick.put_in_pack()
+        picking_pick.move_line_ids.filtered(lambda ml: ml.product_id == self.productB and ml.qty_done == 0.0).qty_done = 2.0
+        second_pack = picking_pick.put_in_pack()
+        picking_pick.button_validate()
+        self.assertEqual(len(first_pack.quant_ids), 2)
+        self.assertEqual(len(second_pack.quant_ids), 1)
+        picking_pack.action_assign()
+        self.assertEqual(len(picking_pack.package_ids), 2)
+        output_sub_location = self.env['stock.location'].create({'name': 'sub1', 'location_id': self.output_location})
+        picking_pack.package_ids.with_context({'picking_id': picking_pack.id, 'destination_location': output_sub_location.id}).action_toggle_processed()
+        self.assertEqual(picking_pack.move_line_ids[:1].location_dest_id, output_sub_location)
+        picking_pack.button_validate()
+        self.assertEqual(first_pack.location_id, output_sub_location)
+        self.assertEqual(second_pack.location_id, output_sub_location)
+        picking_ship.action_assign()
+        self.assertEqual(picking_ship.move_line_ids[:1].location_id, output_sub_location)
+        for move_line in picking_ship.move_line_ids:
+            move_line.qty_done = move_line.product_uom_qty
+        picking_ship.button_validate()
```

Comment:
please assert
- no productA and productB in any location after the last step
- that the quant_ids of the pack in customer are correct